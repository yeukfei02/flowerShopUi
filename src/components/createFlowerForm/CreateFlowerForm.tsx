import React, { useState, useEffect } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import { DropzoneArea } from 'material-ui-dropzone';
import Select from 'react-select';
import axios from 'axios';

import CustomSnackbar from '../customSnackbar/CustomSnackbar';

const ROOT_URL = `https://flower-shop-api.herokuapp.com/api`;

const useStyles = makeStyles({
  root: {
    width: 600,
    margin: 20,
  },
});

const selectStyles = {
  container: (base: any, state: any) => ({
    ...base,
    opacity: state.isDisabled ? '.5' : '1',
    backgroundColor: 'transparent',
    zIndex: '999',
  }),
};

const getShopList = async () => {
  let result = null;

  const response = await axios.get(`${ROOT_URL}/shop`);
  if (response && response.status === 200) {
    if (response.data) {
      const shopList = response.data.shops.map((item: any, i: number) => {
        const obj = {
          value: item.shopId,
          label: item.shopName,
        };
        return obj;
      });
      result = shopList;
    }
  }

  return result;
};

function CreateFlowerForm() {
  const classes = useStyles();

  const [shopList, setShopList] = useState<any[]>([]);

  const [image, setImage] = useState<string>('');
  const [flowerName, setFlowerName] = useState<string>('');
  const [color, setColor] = useState<string>('');
  const [flowerType, setFlowerType] = useState<string>('');
  const [price, setPrice] = useState<number>(0);
  const [occasion, setOccasion] = useState<string>('');
  const [shop, setShop] = useState<any>({ value: 0, label: 'Select shop' });

  const [snackBarStatus, setSnackBarStatus] = useState<string>('');
  const [message, setMessage] = useState<string>('');

  useEffect(() => {
    getShopList()
      .then((result: any) => {
        setShopList(result);
      })
      .catch((e: any) => {
        console.log('error = ', e.message);
      });
  }, []);

  const getBase64 = (file: any, cb: any) => {
    const reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onload = () => {
      cb(reader.result);
    };
    reader.onerror = (error) => {
      console.log('error = ', error);
    };
  };

  const handleFilesUpload = (files: any[]) => {
    if (files && files.length === 1) {
      getBase64(files[0], (imageBase64String: string) => {
        if (imageBase64String) {
          setImage(imageBase64String);
        }
      });
    }
  };

  const handleFlowerNameChange = (e: any) => {
    setFlowerName(e.target.value);
  };

  const handleColorChange = (e: any) => {
    setColor(e.target.value);
  };

  const handleFlowerTypeChange = (e: any) => {
    setFlowerType(e.target.value);
  };

  const handlePriceChange = (e: any) => {
    setPrice(parseFloat(e.target.value));
  };

  const handleOccasionChange = (e: any) => {
    setOccasion(e.target.value);
  };

  const handleShopChange = (selectedShop: any) => {
    if (selectedShop) {
      setShop(selectedShop);
    }
  };

  const createFlower = async (
    image: string,
    flowerName: string,
    color: string,
    flowerType: string,
    price: number,
    occasion: string,
    shop: any,
  ) => {
    const response = await axios.post(
      `${ROOT_URL}/flower/create-flower`,
      {
        image: image,
        flowerName: flowerName,
        color: color,
        flowerType: flowerType,
        price: price,
        occasion: occasion,
        shopId: shop.value,
      },
      {
        headers: {
          'Content-type': 'application/json',
        },
      },
    );
    if (response && response.status === 201) {
      setSnackBarStatus('success');
      setMessage('create flower success');
    } else {
      setSnackBarStatus('error');
      setMessage('create flower error');
    }
  };

  const handleCreateFlower = () => {
    if (image && flowerName && color && flowerType && price && occasion && shop) {
      createFlower(image, flowerName, color, flowerType, price, occasion, shop);
      setSnackBarStatus('');
      setMessage('');
    } else {
      setSnackBarStatus('error');
      setMessage('please enter all fields');
    }
  };

  return (
    <div className="d-flex justify-content-center">
      <Card className={classes.root}>
        <CardContent>
          <Typography variant="h5" gutterBottom>
            Create flower
          </Typography>
          <div className="mt-3 mb-2">
            <DropzoneArea
              acceptedFiles={['image/*']}
              dropzoneText={'Drag and drop an image here or click'}
              filesLimit={1}
              maxFileSize={500000}
              onChange={handleFilesUpload}
              alertSnackbarProps={{
                anchorOrigin: {
                  horizontal: 'center',
                  vertical: 'bottom',
                },
              }}
            />
          </div>
          <TextField
            className="w-100 my-2"
            label="Flower name"
            variant="outlined"
            onChange={(e) => handleFlowerNameChange(e)}
          />
          <TextField className="w-100 my-2" label="Color" variant="outlined" onChange={(e) => handleColorChange(e)} />
          <TextField
            className="w-100 my-2"
            label="Flower type"
            variant="outlined"
            onChange={(e) => handleFlowerTypeChange(e)}
          />
          <TextField
            className="w-100 my-2"
            type="number"
            label="Price"
            variant="outlined"
            onChange={(e) => handlePriceChange(e)}
          />
          <TextField
            className="w-100 my-2"
            label="Occasion"
            variant="outlined"
            onChange={(e) => handleOccasionChange(e)}
          />
          <Select
            className="w-100 my-2"
            styles={selectStyles}
            placeholder={'Select shop'}
            value={shop}
            onChange={handleShopChange}
            options={shopList}
            isClearable={true}
          />
          <Button className="w-100 my-2" variant="contained" size="large" color="primary" onClick={handleCreateFlower}>
            Create flower
          </Button>
        </CardContent>
        <CustomSnackbar snackBarStatus={snackBarStatus} message={message} />
      </Card>
    </div>
  );
}

export default CreateFlowerForm;
