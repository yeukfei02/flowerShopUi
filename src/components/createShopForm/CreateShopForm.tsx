import React, { useState, useEffect } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import axios from 'axios';

const ROOT_URL = `https://flower-shop-api.herokuapp.com/api`;

const useStyles = makeStyles({
  root: {
    width: 600,
    margin: 20
  },
});

function CreateShopForm() {
  const classes = useStyles();

  const [shopName, setShopName] = useState<string>('');
  const [phone, setPhone] = useState<string>('');
  const [address, setAddress] = useState<string>('');

  const handleShopNameChange = (e: any) => {
    if (e.target.value)
      setShopName(e.target.value);
  }

  const handlePhoneChange = (e: any) => {
    if (e.target.value)
      setPhone(e.target.value);
  }

  const handleAddressChange = (e: any) => {
    if (e.target.value)
      setAddress(e.target.value);
  }

  const handleCreateShop = () => {
    if (shopName && phone && address)
      createShop(shopName, phone, address);
  }

  const createShop = async (shopName: string, phone: string, address: string) => {
    const response = await axios.post(`${ROOT_URL}/shop/create-shop`,
      {
        shopName: shopName,
        phone: phone,
        address: address,
      }
    );
    console.log("response = ", response);
  }

  return (
    <div className="d-flex justify-content-center my-4">
      <Card className={classes.root}>
        <CardContent>
          <Typography variant="h5" gutterBottom>
            Create shop
          </Typography>
          <TextField className="w-100 my-2" label="Shop name" variant="outlined" onChange={(e) => handleShopNameChange(e)} />
          <TextField className="w-100 my-2" label="Phone" variant="outlined" onChange={(e) => handlePhoneChange(e)} />
          <TextField className="w-100 my-2" label="Address" variant="outlined" onChange={(e) => handleAddressChange(e)} />
          <Button className="w-100 my-2" variant="contained" size="large" color="primary" onClick={handleCreateShop}>Create shop</Button>
        </CardContent>
      </Card>
    </div>
  );
}

export default CreateShopForm;
