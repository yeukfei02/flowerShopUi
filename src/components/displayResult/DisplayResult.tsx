import React, { useState } from 'react';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import CardMedia from '@material-ui/core/CardMedia';
import HighlightOffIcon from '@material-ui/icons/HighlightOff';
import { useHistory } from 'react-router-dom';
import axios from 'axios';
import _ from 'lodash';

import notFoundImage from '../../images/not-found.png';
import CustomSnackBar from '../customSnackbar/CustomSnackbar';

const ROOT_URL = `https://flower-shop-api.herokuapp.com/api`;

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      flexGrow: 1,
      margin: 20
    },
    paper: {
      padding: theme.spacing(2),
      color: theme.palette.text.primary,
    },
    media: {
      height: 230,
    },
  }),
);

function DisplayResult(props: any) {
  const classes = useStyles();
  const history = useHistory();

  const [snackBarStatus, setSnackBarStatus] = useState<string>('');
  const [message, setMessage] = useState<string>('');

  const getShopById = async (id: number) => {
    let result = '';

    const response = await axios.get(`${ROOT_URL}/shop/${id}`);
    if (response && response.status === 200) {
      if (response.data) {
        if (response.data.shop) {
          result = response.data.shop.shopName;
        }
      }
    }

    return result;
  }

  const deleteShopById = async (id: string) => {
    try {
      const response = await axios.delete(`${ROOT_URL}/shop/${id}`);
      if (response && response.status === 200) {
        setSnackBarStatus('success');
        setMessage(`delete shop by id: ${id}`);
        setTimeout(() => {
          props.fetchData();
        }, 1000);
      }
    } catch (e) {
      console.log("error = ", e.message);
      setSnackBarStatus('error');
      setMessage(`delete shop by id error, please delete flower first`);
    }
  }

  const deleteFlowerById = async (id: string) => {
    try {
      const response = await axios.delete(`${ROOT_URL}/flower/${id}`);
      if (response && response.status === 200) {
        setSnackBarStatus('success');
        setMessage(`delete flower by id: ${id}`);
        setTimeout(() => {
          props.fetchData();
        }, 1000);
      }
    } catch (e) {
      console.log("error = ", e.message);
      setSnackBarStatus('error');
      setMessage(`delete flower by id error`);
    }
  }

  const handleShopDeleteById = (shopId: string) => {
    if (shopId) {
      deleteShopById(shopId);
    }
  }

  const handleFlowerDeleteById = (flowerId: string) => {
    if (flowerId) {
      deleteFlowerById(flowerId);
    }
  }

  const handleGoShopDetails = (shopId: string) => {
    history.push(`/shop/${shopId}`);
  }

  const handleGoFlowerDetails = (flowerId: string) => {
    history.push(`/flower/${flowerId}`);
  }

  const renderShopImage = (item: any) => {
    let cardMedia = (
      <CardMedia
        className={classes.media}
        style={{ cursor: 'pointer' }}
        image={notFoundImage}
        onClick={() => handleGoShopDetails(item.shopId)}
      />
    );

    if (!_.isEqual(item.image, "null")) {
      cardMedia = (
        <CardMedia
          className={classes.media}
          style={{ cursor: 'pointer' }}
          image={item.image}
          onClick={() => handleGoShopDetails(item.shopId)}
        />
      );
    }

    return cardMedia;
  }

  const renderFlowerImage = (item: any) => {
    let cardMedia = (
      <CardMedia
        className={classes.media}
        style={{ cursor: 'pointer' }}
        image={notFoundImage}
        onClick={() => handleGoFlowerDetails(item.flowerId)}
      />
    );

    if (!_.isEqual(item.image, "null")) {
      cardMedia = (
        <CardMedia
          className={classes.media}
          style={{ cursor: 'pointer' }}
          image={item.image}
          onClick={() => handleGoFlowerDetails(item.flowerId)}
        />
      );
    }

    return cardMedia;
  }

  const renderItem = () => {
    let results = null;

    if (props.resultList) {
      if (props.searchValue === 'shop') {
        results = props.resultList.map((item: any, i: number) => {
          return (
            <Grid key={i} item xs={12} sm={4}>
              <Paper className={classes.paper}>
                <div className="d-flex justify-content-end my-2" style={{ cursor: 'pointer' }}>
                  <HighlightOffIcon fontSize="large" onClick={() => handleShopDeleteById(item.shopId)} />
                </div>
                {renderShopImage(item)}
                <Typography className="mt-2" variant="h6" gutterBottom>
                  Shop name: {item.shopName}
                </Typography>
                <Typography variant="h6" gutterBottom>
                  Phone: {item.phone}
                </Typography>
                <Typography variant="h6" gutterBottom>
                  Address: {item.address}
                </Typography>
              </Paper>
            </Grid>
          );
        });
      } else if (props.searchValue === 'flower') {
        results = props.resultList.map((item: any, i: number) => {
          return (
            <Grid key={i} item xs={12} sm={4}>
              <Paper className={classes.paper}>
                <div className="d-flex justify-content-end my-2" style={{ cursor: 'pointer' }}>
                  <HighlightOffIcon fontSize="large" onClick={() => handleFlowerDeleteById(item.flowerId)} />
                </div>
                {renderFlowerImage(item)}
                <Typography className="mt-2" variant="h6" gutterBottom>
                  Flower name: {item.flowerName}
                </Typography>
                <Typography variant="h6" gutterBottom>
                  Color: {item.color}
                </Typography>
                <Typography variant="h6" gutterBottom>
                  Flower type: {item.flowerType}
                </Typography>
                <Typography variant="h6" gutterBottom>
                  Price: {item.price}
                </Typography>
                <Typography variant="h6" gutterBottom>
                  Occasion: {item.occasion}
                </Typography>
                <Typography variant="h6" gutterBottom>
                  Shop: {item.shopId}
                </Typography>
              </Paper>
            </Grid>
          );
        });
      }
    }

    return results;
  }

  return (
    <div className={classes.root}>
      <Grid container spacing={3}>
        {renderItem()}
      </Grid>
      <CustomSnackBar snackBarStatus={snackBarStatus} message={message} />
    </div>
  );
}

export default DisplayResult;
