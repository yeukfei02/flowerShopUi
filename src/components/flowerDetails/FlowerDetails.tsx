import React, { useState, useEffect } from 'react';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import Select from 'react-select';
import { DropzoneArea } from 'material-ui-dropzone';
import { withRouter } from 'react-router';
import { useHistory } from 'react-router-dom';
import axios from 'axios';
import _ from 'lodash';

import notFoundImage from '../../images/not-found.png';
import CustomAppBar from '../customAppBar/CustomAppBar';
import CustomSnackbar from '../customSnackbar/CustomSnackbar';

const ROOT_URL = `https://flower-shop-api.herokuapp.com/api`;

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      flexGrow: 1,
    },
    card: {
      width: 600,
      margin: 20,
    },
    media: {
      height: 230,
    },
  }),
);

const selectStyles = {
  container: (base: any, state: any) => ({
    ...base,
    opacity: state.isDisabled ? '.5' : '1',
    backgroundColor: 'transparent',
    zIndex: '999',
  }),
};

const getShopList = async () => {
  let result = null;

  const response = await axios.get(`${ROOT_URL}/shop`);
  if (response && response.status === 200) {
    if (response.data) {
      const shopList = response.data.shops.map((item: any, i: number) => {
        const obj = {
          value: item.shopId,
          label: item.shopName,
        };
        return obj;
      });
      result = shopList;
    }
  }

  return result;
};

const getFlowerById = async (id: number) => {
  let result = null;

  const response = await axios.get(`${ROOT_URL}/flower/${id}`);
  if (response && response.status === 200) {
    if (response.data && response.data.flower) {
      result = response.data.flower;
    }
  }

  return result;
};

function FlowerDetails(props: any) {
  const classes = useStyles();
  const history = useHistory();

  const [shopList, setShopList] = useState<any[]>([]);

  const [image, setImage] = useState<string>('');
  const [flowerName, setFlowerName] = useState<string>('');
  const [color, setColor] = useState<string>('');
  const [flowerType, setFlowerType] = useState<string>('');
  const [price, setPrice] = useState<number>(0);
  const [occasion, setOccasion] = useState<string>('');
  const [shop, setShop] = useState<any>({ value: 0, label: 'Select shop' });

  const [snackBarStatus, setSnackBarStatus] = useState<string>('');
  const [message, setMessage] = useState<string>('');

  const id = props.match.params.id;

  useEffect(() => {
    getShopList()
      .then((result: any) => {
        setShopList(result);
      })
      .catch((e: any) => {
        console.log('error = ', e.message);
      });
  }, []);

  useEffect(() => {
    if (id) {
      getFlowerById(id)
        .then((result: any) => {
          setImage(result.image);
          setFlowerName(result.flowerName);
          setColor(result.color);
          setFlowerType(result.flowerType);
          setPrice(result.price);
          setOccasion(result.occasion);

          const shopObj = {
            value: result.shopId,
            label: result.shop.shopName,
          };
          setShop(shopObj);
        })
        .catch((e: any) => {
          console.log('error = ', e.message);
        });
    }
  }, [id]);

  const renderFlowerImage = () => {
    let cardMedia = <CardMedia className={classes.media} image={notFoundImage} />;

    if (!_.isEqual(image, 'null')) {
      cardMedia = <CardMedia className={classes.media} image={image} />;
    }

    return cardMedia;
  };

  const getBase64 = (file: any, cb: any) => {
    const reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onload = () => {
      cb(reader.result);
    };
    reader.onerror = (error) => {
      console.log('error = ', error);
    };
  };

  const handleFilesUpload = (files: any[]) => {
    if (files && files.length === 1) {
      getBase64(files[0], (imageBase64String: string) => {
        if (imageBase64String) {
          setImage(imageBase64String);
        }
      });
    }
  };

  const handleFlowerNameChange = (e: any) => {
    setFlowerName(e.target.value);
  };

  const handleColorChange = (e: any) => {
    setColor(e.target.value);
  };

  const handleFlowerTypeChange = (e: any) => {
    setFlowerType(e.target.value);
  };

  const handlePriceChange = (e: any) => {
    setPrice(parseFloat(e.target.value));
  };

  const handleOccasionChange = (e: any) => {
    setOccasion(e.target.value);
  };

  const handleShopChange = (selectedShop: any) => {
    if (selectedShop) {
      setShop(selectedShop);
    }
  };

  const updateFlower = async (
    id: string,
    image: string,
    flowerName: string,
    color: string,
    flowerType: string,
    price: number,
    occasion: string,
    shop: any,
  ) => {
    const response = await axios.patch(
      `${ROOT_URL}/flower/${id}`,
      {
        image: image,
        flowerName: flowerName,
        color: color,
        flowerType: flowerType,
        price: price,
        occasion: occasion,
        shopId: shop.value,
      },
      {
        headers: {
          'Content-type': 'application/json',
        },
      },
    );
    if (response && response.status === 200) {
      setSnackBarStatus('success');
      setMessage('update flower success');
    } else {
      setSnackBarStatus('error');
      setMessage('update flower error');
    }
  };

  const handleUpdateFlower = () => {
    if (image && flowerName && color && flowerType && price && occasion && shop) {
      updateFlower(id, image, flowerName, color, flowerType, price, occasion, shop);
      setSnackBarStatus('');
      setMessage('');
    }
  };

  const handleBack = () => {
    history.push(`/`);
  };

  return (
    <div className={classes.root}>
      <CustomAppBar />
      <div className="d-flex justify-content-center">
        <Card className={classes.card}>
          <CardContent>
            <Typography variant="h5" gutterBottom>
              Flower details
            </Typography>
            {renderFlowerImage()}
            <div className="mt-3 mb-2">
              <DropzoneArea
                acceptedFiles={['image/*']}
                dropzoneText={'Drag and drop an image here or click'}
                filesLimit={1}
                maxFileSize={500000}
                onChange={handleFilesUpload}
                alertSnackbarProps={{
                  anchorOrigin: {
                    horizontal: 'center',
                    vertical: 'bottom',
                  },
                }}
              />
            </div>
            <TextField
              className="w-100 my-2"
              value={flowerName}
              label="Flower name"
              variant="outlined"
              onChange={(e) => handleFlowerNameChange(e)}
            />
            <TextField
              className="w-100 my-2"
              value={color}
              label="Color"
              variant="outlined"
              onChange={(e) => handleColorChange(e)}
            />
            <TextField
              className="w-100 my-2"
              value={flowerType}
              label="Flower type"
              variant="outlined"
              onChange={(e) => handleFlowerTypeChange(e)}
            />
            <TextField
              className="w-100 my-2"
              value={price}
              type="number"
              label="Price"
              variant="outlined"
              onChange={(e) => handlePriceChange(e)}
            />
            <TextField
              className="w-100 my-2"
              value={occasion}
              label="Occasion"
              variant="outlined"
              onChange={(e) => handleOccasionChange(e)}
            />
            <Select
              className="w-100 my-2"
              styles={selectStyles}
              placeholder={'Select shop'}
              value={shop}
              onChange={handleShopChange}
              options={shopList}
              isClearable={true}
            />
            <Button
              className="w-100 my-2"
              variant="contained"
              size="large"
              color="primary"
              onClick={handleUpdateFlower}
            >
              Update flower
            </Button>
            <Button className="w-100 my-2" variant="contained" size="large" color="secondary" onClick={handleBack}>
              Back
            </Button>
          </CardContent>
          <CustomSnackbar snackBarStatus={snackBarStatus} message={message} />
        </Card>
      </div>
    </div>
  );
}

export default withRouter(FlowerDetails);
