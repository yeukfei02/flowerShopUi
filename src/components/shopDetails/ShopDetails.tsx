import React, { useState, useEffect } from 'react';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import { DropzoneArea } from 'material-ui-dropzone';
import { withRouter } from 'react-router';
import { useHistory } from 'react-router-dom';
import axios from 'axios';
import _ from 'lodash';

import notFoundImage from '../../images/not-found.png';
import CustomAppBar from '../customAppBar/CustomAppBar';
import CustomSnackbar from '../customSnackbar/CustomSnackbar';

const ROOT_URL = `https://flower-shop-api.herokuapp.com/api`;

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      flexGrow: 1,
    },
    card: {
      width: 600,
      margin: 20,
    },
    media: {
      height: 230,
    },
  }),
);

const getShopById = async (id: number) => {
  let result = null;

  const response = await axios.get(`${ROOT_URL}/shop/${id}`);
  if (response && response.status === 200) {
    if (response.data && response.data.shop) {
      result = response.data.shop;
    }
  }

  return result;
};

function ShopDetails(props: any) {
  const classes = useStyles();
  const history = useHistory();

  const [image, setImage] = useState<string>('');
  const [shopName, setShopName] = useState<string>('');
  const [phone, setPhone] = useState<string>('');
  const [address, setAddress] = useState<string>('');

  const [snackBarStatus, setSnackBarStatus] = useState<string>('');
  const [message, setMessage] = useState<string>('');

  const id = props.match.params.id;

  useEffect(() => {
    if (id) {
      getShopById(id)
        .then((result: any) => {
          setImage(result.image);
          setShopName(result.shopName);
          setPhone(result.phone);
          setAddress(result.address);
        })
        .catch((e: any) => {
          console.log('error = ', e.message);
        });
    }
  }, [id]);

  const renderShopImage = () => {
    let cardMedia = <CardMedia className={classes.media} style={{ cursor: 'pointer' }} image={notFoundImage} />;

    if (!_.isEqual(image, 'null')) {
      cardMedia = <CardMedia className={classes.media} image={image} />;
    }

    return cardMedia;
  };

  const getBase64 = (file: any, cb: any) => {
    const reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onload = () => {
      cb(reader.result);
    };
    reader.onerror = (error) => {
      console.log('error = ', error);
    };
  };

  const handleFilesUpload = (files: any[]) => {
    if (files && files.length === 1) {
      getBase64(files[0], (imageBase64String: string) => {
        if (imageBase64String) {
          setImage(imageBase64String);
        }
      });
    }
  };

  const handleShopNameChange = (e: any) => {
    setShopName(e.target.value);
  };

  const handlePhoneChange = (e: any) => {
    setPhone(e.target.value);
  };

  const handleAddressChange = (e: any) => {
    setAddress(e.target.value);
  };

  const updateShop = async (id: string, image: string, shopName: string, phone: string, address: string) => {
    const response = await axios.patch(
      `${ROOT_URL}/shop/${id}`,
      {
        image: image,
        shopName: shopName,
        phone: phone,
        address: address,
      },
      {
        headers: {
          'Content-type': 'application/json',
        },
      },
    );
    if (response && response.status === 200) {
      setSnackBarStatus('success');
      setMessage('update shop success');
    } else {
      setSnackBarStatus('error');
      setMessage('update shop error');
    }
  };

  const handleUpdateShop = () => {
    if (image && shopName && phone && address) {
      updateShop(id, image, shopName, phone, address);
      setSnackBarStatus('');
      setMessage('');
    }
  };

  const handleBack = () => {
    history.push(`/`);
  };

  return (
    <div className={classes.root}>
      <CustomAppBar />
      <div className="d-flex justify-content-center">
        <Card className={classes.card}>
          <CardContent>
            <Typography variant="h5" gutterBottom>
              Shop details
            </Typography>
            {renderShopImage()}
            <div className="mt-3 mb-2">
              <DropzoneArea
                acceptedFiles={['image/*']}
                dropzoneText={'Drag and drop an image here or click'}
                filesLimit={1}
                maxFileSize={500000}
                onChange={handleFilesUpload}
                alertSnackbarProps={{
                  anchorOrigin: {
                    horizontal: 'center',
                    vertical: 'bottom',
                  },
                }}
              />
            </div>
            <TextField
              className="w-100 my-2"
              value={shopName}
              label="Shop name"
              variant="outlined"
              onChange={(e) => handleShopNameChange(e)}
            />
            <TextField
              className="w-100 my-2"
              value={phone}
              label="Phone"
              variant="outlined"
              onChange={(e) => handlePhoneChange(e)}
            />
            <TextField
              className="w-100 my-2"
              value={address}
              label="Address"
              variant="outlined"
              onChange={(e) => handleAddressChange(e)}
            />
            <Button className="w-100 my-2" variant="contained" size="large" color="primary" onClick={handleUpdateShop}>
              Update shop
            </Button>
            <Button className="w-100 my-2" variant="contained" size="large" color="secondary" onClick={handleBack}>
              Back
            </Button>
          </CardContent>
          <CustomSnackbar snackBarStatus={snackBarStatus} message={message} />
        </Card>
      </div>
    </div>
  );
}

export default withRouter(ShopDetails);
