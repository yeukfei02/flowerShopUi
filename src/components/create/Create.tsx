import React, { useState } from 'react';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';

import CustomAppBar from '../customAppBar/CustomAppBar';
import CreateShopForm from '../createShopForm/CreateShopForm';
import CreateFlowerForm from '../createFlowerForm/CreateFlowerForm';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      flexGrow: 1,
    },
    card: {
      width: 600,
      margin: 20,
    },
  }),
);

function Create(): JSX.Element {
  const classes = useStyles();

  const [createValue, setCreateValue] = useState<string>('shop');

  const handleCreateShop = (): void => {
    setCreateValue('shop');
  };

  const handleCreateFlower = (): void => {
    setCreateValue('flower');
  };

  const renderCreateForm = (): JSX.Element | null => {
    let createForm = null;

    if (createValue === 'shop') {
      createForm = <CreateShopForm />;
    } else if (createValue === 'flower') {
      createForm = <CreateFlowerForm />;
    }

    return createForm;
  };

  return (
    <div className={classes.root}>
      <CustomAppBar />
      <div className="d-flex justify-content-center my-4">
        <Card className={classes.card}>
          <CardContent>
            <Grid container spacing={3}>
              <Grid item xs={12} sm={6}>
                <Button className="w-100" variant="contained" size="large" color="primary" onClick={handleCreateShop}>
                  Create shop
                </Button>
              </Grid>
              <Grid item xs={12} sm={6}>
                <Button className="w-100" variant="contained" size="large" color="primary" onClick={handleCreateFlower}>
                  Create flower
                </Button>
              </Grid>
            </Grid>
          </CardContent>
          {renderCreateForm()}
        </Card>
      </div>
    </div>
  );
}

export default Create;
