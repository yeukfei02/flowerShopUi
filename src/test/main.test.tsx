import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, mount } from 'enzyme';

import Search from '../components/search/Search';
import Create from '../components/create/Create';
import ShopDetails from '../components/shopDetails/ShopDetails';
import FlowerDetails from '../components/flowerDetails/FlowerDetails';

import CustomAppBar from '../components/customAppBar/CustomAppBar';
import CustomSnackBar from '../components/customSnackBar/CustomSnackBar';
import CreateShopForm from '../components/createShopForm/CreateShopForm';
import CreateFlowerForm from '../components/createFlowerForm/CreateFlowerForm';
import DisplayResult from '../components/displayResult/DisplayResult';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

describe('main.test', () => {
  describe('render test', () => {
    it('Search', () => {
      const wrapper = shallow(<Search />);
      expect(wrapper).toMatchSnapshot();
    });

    it('Create', () => {
      const wrapper = shallow(<Create />);
      expect(wrapper).toMatchSnapshot();
    });

    it('ShopDetails', () => {
      const wrapper = shallow(<ShopDetails />);
      expect(wrapper).toMatchSnapshot();
    });

    it('FlowerDetails', () => {
      const wrapper = shallow(<FlowerDetails />);
      expect(wrapper).toMatchSnapshot();
    });

    it('CustomAppBar', () => {
      const wrapper = shallow(<CustomAppBar />);
      expect(wrapper).toMatchSnapshot();
    });

    it('CustomSnackBar', () => {
      const wrapper = shallow(<CustomSnackBar />);
      expect(wrapper).toMatchSnapshot();
    });

    it('CreateShopForm', () => {
      const wrapper = shallow(<CreateShopForm />);
      expect(wrapper).toMatchSnapshot();
    });

    it('CreateFlowerForm', () => {
      const wrapper = shallow(<CreateFlowerForm />);
      expect(wrapper).toMatchSnapshot();
    });

    it('DisplayResult', () => {
      const wrapper = shallow(<DisplayResult />);
      expect(wrapper).toMatchSnapshot();
    });
  });
});
