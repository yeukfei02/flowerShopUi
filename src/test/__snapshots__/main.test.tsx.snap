// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main.test render test Create 1`] = `ShallowWrapper {}`;

exports[`main.test render test CreateFlowerForm 1`] = `ShallowWrapper {}`;

exports[`main.test render test CreateShopForm 1`] = `ShallowWrapper {}`;

exports[`main.test render test CustomAppBar 1`] = `ShallowWrapper {}`;

exports[`main.test render test CustomSnackBar 1`] = `ShallowWrapper {}`;

exports[`main.test render test DisplayResult 1`] = `ShallowWrapper {}`;

exports[`main.test render test FlowerDetails 1`] = `ShallowWrapper {}`;

exports[`main.test render test Search 1`] = `ShallowWrapper {}`;

exports[`main.test render test ShopDetails 1`] = `ShallowWrapper {}`;
